File: web/providers/kick.js
// Kick public clips client. No auth required, but cross-origin browser calls
// may be CORS-blocked depending on Kick's policy at the time. Mirrors
// providers/kick.py. If CORS fails, the only fix is a small proxy.

const API = "https://kick.com/api/v2";

function pickVideoUrl(c) {
  // Try every reasonable field — Kick has shifted shapes over the years.
  const candidates = [
    c.video_url,
    c.clip_url_mp4,
    c.mp4_url,
    c.download_url,
    c.media_url,
    c.stream_url,
    c.video?.url,
    c.video?.src,
    c.media?.url,
    c.clip?.video_url,
    c.urls?.mp4,
  ];
  for (const url of candidates) {
    if (typeof url === "string" && url) return url;
  }
  return "";
}

function toClip(c, channelSlug) {
  const id = String(c.id || "");
  const channel = c.channel || {};
  const fallback = `https://kick.com/${channelSlug}/clips/${id}`;
  return {
    provider: "kick",
    id,
    title: c.title || "",
    viewCount: Number(c.view_count || c.views || 0),
    durationSec: Number(c.duration || 0),
    thumbnailUrl: c.thumbnail_url || "",
    embedUrl: fallback,
    videoUrl: pickVideoUrl(c),
    url: c.clip_url || fallback,
    broadcasterName: channel.username || channel.slug || channelSlug,
    createdAt: c.created_at || "",
    _raw: c, // dev-only: kept on the object so we can log it once
  };
}

const MAX_PAGES = 10; // safety net: ~200 clips max

export async function getTopClips(channel, win, count = 50) {
  const target = Math.max(1, count);
  const all = [];
  let cursor = null;
  let pages = 0;

  while (all.length < target && pages < MAX_PAGES) {
    const url = new URL(`${API}/channels/${encodeURIComponent(channel)}/clips`);
    url.searchParams.set("sort", "view");
    url.searchParams.set("time", win);
    if (cursor) url.searchParams.set("cursor", cursor);

    let resp;
    try {
      resp = await fetch(url, { credentials: "omit" });
    } catch (err) {
      throw new Error(
        `Kick fetch failed (likely CORS — Kick blocks browser calls from third-party origins): ${err.message}`
      );
    }
    if (!resp.ok) {
      throw new Error(`Kick ${resp.status}: ${resp.statusText}`);
    }

    const body = await resp.json();
    const raw = body.clips || body.data || [];
    if (!raw.length) break;
    all.push(...raw);
    cursor = body.nextCursor || body.next_cursor || body.cursor || null;
    pages += 1;
    if (!cursor) break;
  }

  // One-off debug: dump the first raw clip + top-level keys so we can see
  // what Kick is actually returning. Strip this once the right field names
  // are pinned down.
  if (all[0]) {
    // eslint-disable-next-line no-console
    console.log("[kick debug] first clip top-level keys:", Object.keys(all[0]));
    // eslint-disable-next-line no-console
    console.log("[kick debug] first clip full payload:", all[0]);
  }

  return all.slice(0, target).map((c) => toClip(c, channel));
}

File: src/clip_review/providers/kick.py
"""Kick public clips provider.

Kick's browser-accessible JSON endpoint at /api/v2 is not officially documented
as a public surface but is what kick.com itself uses. CORS may block direct
browser calls from third-party origins; in that case a small proxy is needed.
This module is the authoritative parser regardless of whether the request
originates in the browser or behind a proxy.
"""

from __future__ import annotations

from types import TracebackType
from typing import Any, Self

import httpx

from clip_review.providers import Clip, Window

API_BASE = "https://kick.com/api/v2"

_WINDOW_PARAM: dict[Window, str] = {"day": "day", "week": "week", "month": "month"}


def _to_clip(raw: dict[str, Any], *, channel_slug: str) -> Clip:
    clip_id = str(raw.get("id", ""))
    channel = raw.get("channel") or {}
    name = str(channel.get("username") or channel.get("slug") or channel_slug)
    fallback_url = f"https://kick.com/{channel_slug}/clips/{clip_id}"
    return Clip(
        provider="kick",
        id=clip_id,
        title=str(raw.get("title", "")),
        view_count=int(raw.get("view_count") or raw.get("views") or 0),
        duration_sec=float(raw.get("duration") or 0.0),
        thumbnail_url=str(raw.get("thumbnail_url") or ""),
        embed_url=fallback_url,
        url=str(raw.get("clip_url") or fallback_url),
        broadcaster_name=name,
        created_at=str(raw.get("created_at", "")),
    )


class KickProvider:
    """Async helper around Kick's public clips endpoint. No auth required."""

    def __init__(self, *, http: httpx.AsyncClient | None = None) -> None:
        self._http = http or httpx.AsyncClient(timeout=10.0)
        self._owns_http = http is None

    async def __aenter__(self) -> Self:
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        if self._owns_http:
            await self._http.aclose()

    async def get_top_clips(self, channel: str, window: Window, count: int = 50) -> list[Clip]:
        params = {"sort": "view", "time": _WINDOW_PARAM[window]}
        resp = await self._http.get(f"{API_BASE}/channels/{channel}/clips", params=params)
        resp.raise_for_status()
        body = resp.json()
        raw_clips = body.get("clips") or body.get("data") or []
        clips = [_to_clip(c, channel_slug=channel) for c in raw_clips]
        return clips[: max(1, count)]
